resteasy-core/src/main/java/org/jboss/resteasy/core/interception/jaxrs/ContainerResponseContextImpl.java

```diff
@@ -125,6 +125,17 @@ public Type getEntityType()
    public void setEntity(Object entity)
    {
       //if (entity != null) logger.info("*** setEntity(Object) " + entity.toString());
+
+      // In case the entity was null, new response is created since the response is "204 No Content"
+      // and as such the response entity would be null even when set to non-null one
+      if (!jaxrsResponse.hasEntity() && entity != null){
+         Response.ResponseBuilder builder = Response.ok(entity);
+         BuiltResponse response = (BuiltResponse)builder.build();
+
+         jaxrsResponse.setStatus(response.getStatus());
+         jaxrsResponse.setReasonPhrase(response.getReasonPhrase());
+      }

```

Comment:
These aren't quite right. For one they should build a new response. Another issue though is it resets the status which seems wrong. While the 204 may seem weird, the filter should handle. If, for example, this was a 404 then this resets it to 200 and gives an okay response with the entity.

One option might be to check the current status with something like:
```java
if (!jaxrsResponse.hasEntity() && entity != null && jaxrsResponse.getStatus() == Response.Status.NO_CONTENT.getStatusCode()) {
   jaxrsResponse.setStatus(Response.Status.OK.getStatusCode());
   jaxrsResponse.setReasonPhrase(Response.Status.OK.getReasonPhrase());
}
```
That said, I think we should remove this. The `ContainerResponseFilter` should be responsible for setting this IMO.